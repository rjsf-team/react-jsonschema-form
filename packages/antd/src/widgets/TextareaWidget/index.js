import React from 'react';

import Input from 'antd/lib/input';

const INPUT_STYLE = {
  width: '100%',
};

const TextareaWidget = ({
  // autofocus,
  disabled,
  formContext,
  id,
  // label,
  onBlur,
  onChange,
  onFocus,
  options,
  placeholder,
  readonly,
  // required,
  // schema,
  value,
}) => {
  const { readonlyAsDisabled = true } = formContext;

  const handleChange = ({ target }) =>
    onChange(target.value === '' ? options.emptyValue : target.value);

  const handleBlur = ({ target }) => onBlur(id, target.value);

  const handleFocus = ({ target }) => onFocus(id, target.value);

  return (
    <Input.TextArea
      disabled={disabled || (readonlyAsDisabled && readonly)}
      id={id}
      name={id}
      onBlur={!readonly ? handleBlur : undefined}
      onChange={!readonly ? handleChange : undefined}
      onFocus={!readonly ? handleFocus : undefined}
      placeholder={placeholder}
      rows={options.rows || 4}
      style={INPUT_STYLE}
      type="textarea"
      value={value}
    />
  );
};

export default TextareaWidget;
