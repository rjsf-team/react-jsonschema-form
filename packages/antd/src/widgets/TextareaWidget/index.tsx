import { ChangeEvent, FocusEvent } from 'react';
import Input from 'antd/lib/input';
import {
  ariaDescribedByIds,
  FormContextType,
  GenericObjectType,
  RJSFSchema,
  StrictRJSFSchema,
  WidgetProps,
} from '@rjsf/utils';

const INPUT_STYLE = {
  width: '100%',
};

/** The `TextareaWidget` is a widget for rendering input fields as textarea.
 *
 * @param props - The `WidgetProps` for this component
 */
export default function TextareaWidget<
  T = any,
  S extends StrictRJSFSchema = RJSFSchema,
  F extends FormContextType = any
>({
  disabled,
  formContext,
  id,
  onBlur,
  onChange,
  onFocus,
  options,
  placeholder,
  readonly,
  value,
}: WidgetProps<T, S, F>) {
  const { readonlyAsDisabled = true } = formContext as GenericObjectType;

  const handleChange = ({ target }: ChangeEvent<HTMLTextAreaElement>) =>
    onChange(target.value === '' ? options.emptyValue : target.value);

  const handleBlur = ({ target }: FocusEvent<HTMLTextAreaElement>) => onBlur(id, target.value);

  const handleFocus = ({ target }: FocusEvent<HTMLTextAreaElement>) => onFocus(id, target.value);

  // Antd's typescript definitions do not contain the following props that are actually necessary and, if provided,
  // they are used, so hacking them in via by spreading `extraProps` on the component to avoid typescript errors
  const extraProps = {
    type: 'textarea',
  };

  return (
    <Input.TextArea
      disabled={disabled || (readonlyAsDisabled && readonly)}
      id={id}
      name={id}
      onBlur={!readonly ? handleBlur : undefined}
      onChange={!readonly ? handleChange : undefined}
      onFocus={!readonly ? handleFocus : undefined}
      placeholder={placeholder}
      rows={options.rows || 4}
      style={INPUT_STYLE}
      value={value}
      {...extraProps}
      aria-describedby={ariaDescribedByIds<T>(id)}
    />
  );
}
