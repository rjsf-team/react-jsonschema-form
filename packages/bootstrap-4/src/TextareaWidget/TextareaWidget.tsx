import React from "react";

import { WidgetProps } from "@rjsf/core";
import FormControl from "react-bootstrap/FormControl";
import InputGroup from "react-bootstrap/InputGroup";

type CustomWidgetProps = WidgetProps & {
  options: any;
};

const TextareaWidget = ({
  id,
  placeholder,
  value,
  required,
  disabled,
  autofocus,
  label,
  readonly,
  onBlur,
  onFocus,
  onChange,
  options,
  schema,
  rawErrors = [],
}: CustomWidgetProps) => {
  const _onChange = ({
    target: { value },
  }: React.ChangeEvent<HTMLTextAreaElement>) =>
    onChange(value === "" ? options.emptyValue : value);
  const _onBlur = ({
    target: { value },
  }: React.FocusEvent<HTMLTextAreaElement>) => onBlur(id, value);
  const _onFocus = ({
    target: { value },
  }: React.FocusEvent<HTMLTextAreaElement>) => onFocus(id, value);

  return (
    <>
      <label htmlFor={id}>
        {label || schema.title}
        {required && (
          <span
            aria-hidden
            className={rawErrors.length > 0 ? "text-danger ml-1" : "ml-1"}>
            &thinsp;{"*"}
          </span>
        )}
      </label>
      <InputGroup>
        <FormControl
          id={id}
          as="textarea"
          placeholder={placeholder}
          disabled={disabled}
          readOnly={readonly}
          value={value}
          required={required}
          autoFocus={autofocus}
          rows={options.rows || 5}
          onChange={_onChange}
          onBlur={_onBlur}
          onFocus={_onFocus}
        />
      </InputGroup>
    </>
  );
};

export default TextareaWidget;
