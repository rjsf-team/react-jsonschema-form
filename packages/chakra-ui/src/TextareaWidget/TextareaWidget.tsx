import { ChangeEvent, FocusEvent } from 'react';
import { FormControl, FormLabel, Textarea } from '@chakra-ui/react';
import {
  ariaDescribedByIds,
  labelValue,
  FormContextType,
  RJSFSchema,
  StrictRJSFSchema,
  WidgetProps,
} from '@rjsf/utils';
import { getChakra } from '../utils';

export default function TextareaWidget<
  T = any,
  S extends StrictRJSFSchema = RJSFSchema,
  F extends FormContextType = any
>({
  id,
  placeholder,
  value,
  label,
  hideLabel,
  disabled,
  autofocus,
  readonly,
  onBlur,
  onFocus,
  onChange,
  options,
  uiSchema,
  required,
  rawErrors,
}: WidgetProps<T, S, F>) {
  const chakraProps = getChakra({ uiSchema });

  const _onChange = ({ target: { value } }: ChangeEvent<HTMLTextAreaElement>) =>
    onChange(value === '' ? options.emptyValue : value);
  const _onBlur = ({ target: { value } }: FocusEvent<HTMLTextAreaElement>) => onBlur(id, value);
  const _onFocus = ({ target: { value } }: FocusEvent<HTMLTextAreaElement>) => onFocus(id, value);

  return (
    <FormControl
      mb={1}
      {...chakraProps}
      isDisabled={disabled || readonly}
      isRequired={required}
      isReadOnly={readonly}
      isInvalid={rawErrors && rawErrors.length > 0}
    >
      {labelValue(<FormLabel htmlFor={id}>{label}</FormLabel>, hideLabel || !label)}
      <Textarea
        id={id}
        name={id}
        value={value ?? ''}
        placeholder={placeholder}
        autoFocus={autofocus}
        onChange={_onChange}
        onBlur={_onBlur}
        onFocus={_onFocus}
        aria-describedby={ariaDescribedByIds<T>(id)}
      />
    </FormControl>
  );
}
