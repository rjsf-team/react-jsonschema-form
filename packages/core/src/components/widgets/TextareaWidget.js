import React from "react";
import PropTypes from "prop-types";

function TextareaWidget(props) {
  const {
    id,
    options,
    placeholder,
    value,
    required,
    disabled,
    readonly,
    autofocus,
    onChange,
    onBlur,
    onFocus,
  } = props;
  const _onChange = ({ target: { value } }) => {
    return onChange(value === "" ? options.emptyValue : value);
  };
  return (
    <textarea
      id={id}
      className="form-control"
      value={value ? value : ""}
      placeholder={placeholder}
      required={required}
      disabled={disabled}
      readOnly={readonly}
      autoFocus={autofocus}
      rows={options.rows}
      onBlur={onBlur && (event => onBlur(id, event.target.value))}
      onFocus={onFocus && (event => onFocus(id, event.target.value))}
      onChange={_onChange}
    />
  );
}

TextareaWidget.defaultProps = {
  autofocus: false,
  options: {},
};

if (process.env.NODE_ENV !== "production") {
  TextareaWidget.propTypes = {
    schema: PropTypes.object.isRequired,
    id: PropTypes.string.isRequired,
    placeholder: PropTypes.string,
    options: PropTypes.shape({
      rows: PropTypes.number,
    }),
    value: PropTypes.string,
    required: PropTypes.bool,
    disabled: PropTypes.bool,
    readonly: PropTypes.bool,
    autofocus: PropTypes.bool,
    onChange: PropTypes.func,
    onBlur: PropTypes.func,
    onFocus: PropTypes.func,
  };
}

export default TextareaWidget;
