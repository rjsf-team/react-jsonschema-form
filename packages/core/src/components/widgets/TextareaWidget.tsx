import { ChangeEvent, FocusEvent, useCallback } from 'react';
import { ariaDescribedByIds, FormContextType, RJSFSchema, StrictRJSFSchema, WidgetProps } from '@rjsf/utils';

/** The `TextareaWidget` is a widget for rendering input fields as textarea.
 *
 * @param props - The `WidgetProps` for this component
 */
function TextareaWidget<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>({
  id,
  options = {},
  placeholder,
  value,
  required,
  disabled,
  readonly,
  autofocus = false,
  onChange,
  onBlur,
  onFocus,
}: WidgetProps<T, S, F>) {
  const handleChange = useCallback(
    ({ target: { value } }: ChangeEvent<HTMLTextAreaElement>) => onChange(value === '' ? options.emptyValue : value),
    [onChange, options.emptyValue]
  );

  const handleBlur = useCallback(
    ({ target: { value } }: FocusEvent<HTMLTextAreaElement>) => onBlur(id, value),
    [onBlur, id]
  );

  const handleFocus = useCallback(
    ({ target: { value } }: FocusEvent<HTMLTextAreaElement>) => onFocus(id, value),
    [id, onFocus]
  );

  return (
    <textarea
      id={id}
      name={id}
      className='form-control'
      value={value ? value : ''}
      placeholder={placeholder}
      required={required}
      disabled={disabled}
      readOnly={readonly}
      autoFocus={autofocus}
      rows={options.rows}
      onBlur={handleBlur}
      onFocus={handleFocus}
      onChange={handleChange}
      aria-describedby={ariaDescribedByIds<T>(id)}
    />
  );
}

TextareaWidget.defaultProps = {
  autofocus: false,
  options: {},
};

export default TextareaWidget;
