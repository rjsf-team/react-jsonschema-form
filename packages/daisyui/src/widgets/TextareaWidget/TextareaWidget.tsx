import { FocusEvent, useCallback } from 'react';
import { WidgetProps, StrictRJSFSchema, RJSFSchema, FormContextType } from '@rjsf/utils';

/** The `TextareaWidget` component renders a multi-line text input with DaisyUI styling
 *
 * Features:
 * - Handles multi-line text input with proper styling
 * - Supports required, disabled, and readonly states
 * - Manages focus and blur events for accessibility
 * - Uses DaisyUI's textarea component for consistent styling
 *
 * @param props - The `WidgetProps` for this component
 */
export default function TextareaWidget<
  T = any,
  S extends StrictRJSFSchema = RJSFSchema,
  F extends FormContextType = any,
>(props: WidgetProps<T, S, F>) {
  const { id, value, required, disabled, readonly, onChange, onFocus, onBlur, options } = props;

  /** Handle focus events
   *
   * @param event - The focus event
   */
  const handleFocus = useCallback(
    (event: FocusEvent<HTMLTextAreaElement>) => {
      if (onFocus) {
        onFocus(id, event.target.value);
      }
    },
    [onFocus, id],
  );

  /** Handle blur events
   *
   * @param event - The blur event
   */
  const handleBlur = useCallback(
    (event: FocusEvent<HTMLTextAreaElement>) => {
      if (onBlur) {
        onBlur(id, event.target.value);
      }
    },
    [onBlur, id],
  );

  /** Handle change events
   *
   * @param event - The change event
   */
  const handleChange = useCallback(
    (event: React.ChangeEvent<HTMLTextAreaElement>) => {
      onChange(event.target.value);
    },
    [onChange],
  );

  // Extract rows and other textarea-specific props from options
  const rows = options?.rows || 5;

  return (
    <div className='form-control'>
      <textarea
        id={id}
        value={value || ''}
        required={required}
        disabled={disabled || readonly}
        readOnly={readonly}
        rows={rows}
        onChange={handleChange}
        onFocus={handleFocus}
        onBlur={handleBlur}
        className='textarea textarea-bordered w-full'
      />
    </div>
  );
}
