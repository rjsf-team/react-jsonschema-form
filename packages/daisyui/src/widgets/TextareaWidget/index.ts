export { default } from './TextareaWidget';
export * from './TextareaWidget';
