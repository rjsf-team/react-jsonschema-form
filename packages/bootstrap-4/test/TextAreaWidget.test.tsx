import React from "react";
import TextareaWidget from "../src/TextareaWidget";
import { makeWidgetMockProps } from "./helpers/createMocks";
import renderer from "react-test-renderer";

describe("TextareaWidget", () => {
  test("simple without errors", () => {
    const tree = renderer
      .create(<TextareaWidget {...makeWidgetMockProps({})} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("simple with errors", () => {
    const tree = renderer
      .create(
        <TextareaWidget
          {...makeWidgetMockProps({ rawErrors: ["Invalid 1", "Invalid 2"] })}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("simple without required", () => {
    const tree = renderer
      .create(<TextareaWidget {...makeWidgetMockProps({ required: false })} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
