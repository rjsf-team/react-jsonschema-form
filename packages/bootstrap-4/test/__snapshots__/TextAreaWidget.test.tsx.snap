// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextareaWidget simple with errors 1`] = `
Array [
  <label
    htmlFor="_id"
  >
    Some simple label
    <span
      aria-hidden={true}
      className="text-danger ml-1"
    >
       
      *
    </span>
  </label>,
  <div
    className="input-group"
  >
    <textarea
      autoFocus={true}
      className="form-control"
      disabled={false}
      id="_id"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      readOnly={true}
      required={true}
      rows={5}
      value="value"
    />
  </div>,
]
`;

exports[`TextareaWidget simple without errors 1`] = `
Array [
  <label
    htmlFor="_id"
  >
    Some simple label
    <span
      aria-hidden={true}
      className="text-danger ml-1"
    >
       
      *
    </span>
  </label>,
  <div
    className="input-group"
  >
    <textarea
      autoFocus={true}
      className="form-control"
      disabled={false}
      id="_id"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      readOnly={true}
      required={true}
      rows={5}
      value="value"
    />
  </div>,
]
`;

exports[`TextareaWidget simple without required 1`] = `
Array [
  <label
    htmlFor="_id"
  >
    Some simple label
  </label>,
  <div
    className="input-group"
  >
    <textarea
      autoFocus={true}
      className="form-control"
      disabled={false}
      id="_id"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      readOnly={true}
      required={false}
      rows={5}
      value="value"
    />
  </div>,
]
`;
